lib/amazonka/src/Amazonka/Auth/SSO.hs

```diff
@@ -0,0 +1,165 @@
+-- Stability   : provisional
+-- Portability : non-portable (GHC extensions)
+--
+-- Authentication using a JWT stored in @~/.aws/sso/cache/{uuid}.json@ as
+-- created by @aws sso login@. For more details, see
+-- <https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-sso.html>.
+module Amazonka.Auth.SSO where
+
+import Amazonka.Auth.Background (fetchAuthInBackground)
+import Amazonka.Auth.ConfigFile (configPathRelative, loadIniFile)
+import Amazonka.Auth.Exception
+import Amazonka.Data.Sensitive
+import Amazonka.Data.Time (ISO8601, Time (..))
+import Amazonka.Env (Env, Env' (..), lookupRegion)
+import Amazonka.HTTP (retryRequest)
+import Amazonka.Lens (throwingM, (^.))
+import Amazonka.Prelude
+import Amazonka.SSO.GetRoleCredentials as SSO
+import qualified Amazonka.SSO.Types as SSO (RoleCredentials (..))
+import Amazonka.Types
+import qualified Control.Exception as Exception
+import Control.Monad.Trans.Resource (runResourceT)
+import Data.Aeson (FromJSON, decodeFileStrict)
+import qualified Data.HashMap.Strict as HashMap
+import Data.Text (pack)
+import Data.Text.Encoding (encodeUtf8)
+import Data.Time.Clock.POSIX (posixSecondsToUTCTime)
+import qualified Network.HTTP.Client as Client
+import qualified System.Directory as Directory
+import qualified System.Environment as Environment
+import System.FilePath.Glob (decompile, globDir1)
+
+-- | Data present in @~/.aws/sso/cache/x.json@
+data Cache = Cache
+  { accessToken :: Text,
+    region :: Region
+  }
+  deriving stock (Generic)
+  deriving anyclass (FromJSON)
+
+-- | Data present in @~/.aws/config@
+data Config = Config
+  { sso_account_id :: Text,
+    sso_role_name :: Text
+  }
+
+fromSSOCache ::
+  ( MonadIO m,
+    Foldable withAuth
+  ) =>
+  Env' withAuth ->
+  m Env
+fromSSOCache env = do
+  auth <- liftIO $ fetchAuthInBackground getCredentials
+  discoverRegion <&> \case
+    Nothing -> env {envAuth = Identity auth}
+    Just r -> env {envRegion = r, envAuth = Identity auth}
+  where
+    getCredentials = do
+      Cache {..} <- readCache
+      Config {..} <- readConfig
+
+      -- Use the SSO region, even if configuring for another
+      let ssoEnv = env {envRegion = region}
+
+      resp <-
+        runResourceT $
+          sendWithoutAuth ssoEnv $
+            SSO.newGetRoleCredentials sso_role_name sso_account_id accessToken
+
+      let mCreds = do
+            rc <- resp ^. SSO.getRoleCredentialsResponse_roleCredentials
+            roleCredentialsToAuthEnv rc
+
+      case mCreds of
+        Nothing -> fail "sso:GetRoleWithCredentials returned no credentials."
+        Just c -> pure c
+
+-- | Send an unsigned request and read the response body
+sendWithoutAuth ::
+  ( MonadResource m,
+    AWSRequest a,
+    Foldable withAuth
+  ) =>
+  Env' withAuth ->
+  a ->
+  m (AWSResponse a)
+sendWithoutAuth env req = do
+  eResponse <- retryRequest env req
+  either (liftIO . Exception.throwIO) (pure . Client.responseBody) eResponse
```

Comment:
I'd like to see this pulled out somewhere and used to DRY up `Amazonka.Auth.STS`' `assumeRoleWithWebIdentity` support. I don't think it needs to be part of the public interface, but maybe `Amazonka.Auth.Internal.sendWithoutAuth` (new module) would be good?